sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -52,76 +91,183 @@ def _get_component_hash(cls, component: Component) -> str:
             object_hash = hashlib.sha256()
             object_hash.update(component._get_anonymous_hash().encode("utf-8"))  # pylint: disable=protected-access
             object_hash.update(component._source_path.encode("utf-8"))  # pylint: disable=protected-access
-            return cls._YAML_SOURCE_PREFIX + object_hash.hexdigest()
+            return _YAML_SOURCE_PREFIX + object_hash.hexdigest()
         # For components without code, like pipeline component, their dependencies have already
         # been resolved before calling this function, so we can use their anonymous hash directly
-        return cls._ANONYMOUS_HASH_PREFIX + component._get_anonymous_hash()  # pylint: disable=protected-access
+        return _ANONYMOUS_HASH_PREFIX + component._get_anonymous_hash()  # pylint: disable=protected-access
 
-    def register_node_to_resolve(self, node: BaseNode):
-        """Register a node with its component to resolve.
+    @staticmethod
+    def _get_on_disk_hash_for_component(component: Component, in_memory_hash: str) -> str:
+        """Get a hash for a component. Will calculate the hash based on the component's code folder if
+        the component has code, otherwise will use the in-memory hash.
```

Comment:
Why the hashing logic is different for in memory and on disc hash?

Reply:
We can't assume that the code folder won't change among dependency resolution.

In in-memory hash calculate, we use path to calculate hash instead of folder content to avoid io operations, which is not safe across 2 `jobs.create_or_update` calls.

Reply:
Will update docstring accordingly.